pkg/typesystem/typesystem_test.go

```diff
@@ -12,6 +12,490 @@ import (
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [folder]`,
+			inputType:     "document",
+			inputRelation: "unknown",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`undefined_relation_in_assignable_type`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document#unknown]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`undefined_computed_userset`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: unknown`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`undefined_tupleset`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: viewer from unknown`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`this_has_entrypoints_through_user`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_has_entrypoints_through_user_wildcard`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document:*]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_has_entrypoints_through_userset`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type org
+				relations
+					define member: [user]
+			type folder
+				relations
+					define parent: [org#member]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_with_two_assignable_types_has_entrypoints_through_first`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type folder
+				relations
+					define parent: [user, folder#parent]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_with_two_assignable_types_has_entrypoints_through_second`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type folder
+				relations
+					define editor: [user]
+					define parent: [folder#parent, folder#editor]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		// TODO fix
+		//`this_has_no_entrypoints_because_type_unknown_is_not_defined`: {
+		//	model: `
+		//	model
+		//		schema 1.1
+		//	type folder
+		//		relations
+		//			define parent: [unknown]`,
+		//	inputType:     "folder",
+		//	inputRelation: "parent",
+		//	expectError:   "undefined type 'unknown'",
+		//},
+		`this_has_no_entrypoints`: {
```

Comment:
Can we also add one more nuanced test, just for good measure, that tests the following scenario:

```
type group
  relations
    define member: [group#member]

type folder
  relations
    define parent: [group#member]
```

* InputType: folder
* InputRelation: parent

The idea here is to test that if we follow a userset through a direct relation and that userset defines no entrypoints, then we should inherit that outcome. Recursively speaking, a type restriction for some input type with no entrypoints should always produce no entrypoints for the target input.

Reply:
added in `this_has_no_entrypoints_through_recursive_userset`